odoo/api.py

```diff
@@ -726,19 +726,20 @@ def reset(cls):
         """
         cls._local.environments = Environments()
 
-    def __new__(cls, cr, uid, context):
+    def __new__(cls, cr, uid, context, real_uid=None):
         assert context is not None
         args = (cr, uid, context)
 
         # if env already exists, return it
         env, envs = None, cls.envs
         for env in envs:
-            if env.args == args:
+            if env.args == args and env.real_uid == real_uid:
                 return env
 
         # otherwise create environment, and add it in the set
         self = object.__new__(cls)
         self.cr, self.uid, self.context = self.args = (cr, uid, frozendict(context))
+        self.real_uid = real_uid
```

Comment:
If Env() is called directly then self.real_uid will be None. Is that what you intended?
Should it be set to self.uid instead? That would make its value always an integer id.

Reply:
The original intention was to not expose the attr and rely on the getter for the update.
Initialising it to the current uid would make the getter superfluous as you mentioned, but we'll need to move the update check to the caller method. No strong opinion.